x-pack/platform/plugins/private/reporting/server/routes/common/scheduled/scheduled_query.ts

```diff
@@ -0,0 +1,290 @@
+import { SearchResponse } from '@elastic/elasticsearch/lib/api/types';
+import { RRule } from '@kbn/rrule';
+import type { ReportingCore } from '../../..';
+import type {
+  ListScheduledReportApiJSON,
+  ReportingUser,
+  ScheduledReportType,
+} from '../../../types';
+import { SCHEDULED_REPORT_SAVED_OBJECT_TYPE } from '../../../saved_objects';
+
+export const MAX_SCHEDULED_REPORT_LIST_SIZE = 100;
+export const DEFAULT_SCHEDULED_REPORT_LIST_SIZE = 10;
+
+const SCHEDULED_REPORT_ID_FIELD = 'scheduled_report_id';
+const CREATED_AT_FIELD = 'created_at';
+const getUsername = (user: ReportingUser) => (user ? user.username : false);
+
+interface ApiResponse {
+  page: number;
+  per_page: number;
+  total: number;
+  data: ListScheduledReportApiJSON[];
+}
+
+const getEmptyApiResponse = (page: number, perPage: number) => ({
+  page,
+  per_page: perPage,
+  total: 0,
+  data: [],
+});
+
+interface BulkOperationError {
+  message: string;
+  status?: number;
+  id: string;
+}
+
+interface BulkDisableResult {
+  scheduled_report_ids: string[];
+  errors: BulkOperationError[];
+  total: number;
+}
+
+export type CreatedAtSearchResponse = SearchResponse<{ created_at: string }>;
+
+export function transformResponse(
+  result: SavedObjectsFindResponse<ScheduledReportType>,
+  lastResponse?: CreatedAtSearchResponse
+): ApiResponse {
+  return {
+    page: result.page,
+    per_page: result.per_page,
+    total: result.total,
+    data: result.saved_objects.map((so) => {
+      const id = so.id;
+      const lastRunForId = (lastResponse?.hits.hits ?? []).find(
+        (hit) => hit.fields?.[SCHEDULED_REPORT_ID_FIELD]?.[0] === id
+      );
+
+      const schedule = so.attributes.schedule;
+      const _rrule = new RRule({
+        ...schedule.rrule,
+        dtstart: new Date(),
+      });
+
+      return {
+        id,
+        created_at: so.attributes.createdAt,
+        created_by: so.attributes.createdBy,
+        enabled: so.attributes.enabled,
+        jobtype: so.attributes.jobType,
+        object_type: so.attributes.meta.objectType,
+        last_run: lastRunForId?._source?.[CREATED_AT_FIELD],
+        next_run: _rrule.after(new Date())?.toISOString(),
+        notification: so.attributes.notification,
+        schedule: so.attributes.schedule,
+        title: so.attributes.title,
+      };
+    }),
+  };
+}
+
+export interface ScheduledQueryFactory {
+  list(
+    req: KibanaRequest,
+    res: KibanaResponseFactory,
+    user: ReportingUser,
+    page: number,
+    size: number
+  ): Promise<ApiResponse>;
+  bulkDisable(
+    logger: Logger,
+    req: KibanaRequest,
+    res: KibanaResponseFactory,
+    ids: string[],
+    user: ReportingUser
+  ): Promise<BulkDisableResult>;
+}
+
+export function scheduledQueryFactory(reportingCore: ReportingCore): ScheduledQueryFactory {
+  return {
+    async list(req, res, user, page = 1, size = DEFAULT_SCHEDULED_REPORT_LIST_SIZE) {
```

Comment:
Feels like we should put each route in a separate file, but perhaps we can plan on doing that when we add the next set of APIs (enable?)